Editor - Move Undo/Redo to the header

geriux opened the pull request:
**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/5889
- https://github.com/WordPress/gutenberg/pull/51766
- https://github.com/wordpress-mobile/WordPress-Android/pull/18705

This PR adds new undo/redo buttons in the navigation bar when the editor is opened, it also updates the existing menu icon.

## Screenshots

iPhone Light mode | iPhone Dark mode
-|-
<kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/4766e0d1-2e13-4e14-bdac-04f02bb1b6aa" width="250" /></kbd> | <kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/44a5274f-75ea-4060-9004-23d6de359489" width="250" /></kbd>

iPhone Landscape Light mode | iPhone Landscape Dark mode
-|-
<kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/37db8e0a-c9e8-408a-a966-94b32f6f5e4b" width="250" /></kbd> | <kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/558a8861-ea44-4ebf-b258-8a36b7afd3c0" width="250" /></kbd>

iPhone RTL |
-|
<kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/a311549a-2ef1-4091-be32-d6194f331597" width="250" /></kbd> | 

iPad Light mode | iPad Dark mode
-|-
<kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/9ce33641-4a0d-49f5-aec9-473e1dd9e806" width="250" /></kbd> | <kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/d9d2e811-af3f-4b01-b964-1ef09591d3a6" width="250" /></kbd>

iPad Landscape Light mode | iPad Landscape Dark mode
-|-
<kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/dd35f962-ec6f-4971-b6ed-c4526cd03447" width="250" /></kbd> | <kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/933c1696-5b70-41a3-92a6-53b9eaab4977" width="250" /></kbd>

## To test

- Install the build
- Create a new post
- Add a few blocks/text
- Tap on the new undo/redo buttons in the navigation bar
- **Expect** the content to change accordingly depending on the button you've tapped

**Bonus testing**: [follow these test cases](https://github.com/wordpress-mobile/test-cases/tree/trunk/test-cases/gutenberg/writing-flow#undo--redo---test-cases).

## Regression Notes
1. Potential unintended areas of impact
Post publishing

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Rely on current tests and manual testing

3. What automated tests I added (or what prevented me from doing so)
Added new tests for testing the new undo/redo buttons in the navigation bar.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] VoiceOver.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [x] Multi-tasking: Split view and Slide over. (iPad)

Commit:
Editor - Move Undo/Redo to the header

Commit:
Post Editor - Update Site icon size and spacing between buttons

Commit:
Editor Bar Manager - Add opacity change to custom buttom on pressed

Commit:
Add UI tests for the Undo/Redo functionality

Commit:
Update Gutenberg ref

Commit:
Remove duplicated UITest button after fixing conflicts

Commit:
Fix border line for the navigation bar for landscape orientation

Commit:
Fix undo/redo buttons for RTL

Commit:
Fix test

tiagomar commented on WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift:
```diff
@@ -236,6 +246,36 @@ public class BlockEditorScreen: ScreenObject {
+        undoButton.tap()
+
+        return try BlockEditorScreen()
+    }
```
WDYT about removing the assertion here ? `undoButton` is already listed as expected element in `public init(app: XCUIApplication = XCUIApplication())`, the test would fail anyways if not hittable, and the extra waiting only adds up to the total execution time.

geriux replied:
Thanks for the suggestion! I've applied the suggested changes in https://github.com/wordpress-mobile/WordPress-iOS/pull/20930/commits/ce5d63754b3a26091176e73640bea72e5720326c

tiagomar commented on WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift:
```diff
@@ -236,6 +246,36 @@ public class BlockEditorScreen: ScreenObject {
+    }
+
+    @discardableResult
+    public func redo() throws -> BlockEditorScreen {
+        XCTAssertTrue(redoButton.waitForIsHittable(timeout: 3))
+        redoButton.tap()
+
+        return try BlockEditorScreen()
+    }
```
Same as in `undo()`.

geriux replied:
Updated in https://github.com/wordpress-mobile/WordPress-iOS/pull/20930/commits/ce5d63754b3a26091176e73640bea72e5720326c

tiagomar commented on WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift:
```diff
@@ -236,6 +246,36 @@ public class BlockEditorScreen: ScreenObject {
+    }
+
+    @discardableResult
+    public func undoIsDisabled() throws -> BlockEditorScreen {
```
```suggestion
    public func verifyUndoIsDisabled() throws -> BlockEditorScreen {
```
This would be just to follow the existing pattern to differentiate action steps from assertions, making it clearer for someone reading the test case.

geriux replied:
Sounds good! I've updated it in https://github.com/wordpress-mobile/WordPress-iOS/pull/20930/commits/9747901ba9f5465abdb836197773818f4988ec31 thanks!

tiagomar commented on WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift:
```diff
@@ -236,6 +246,36 @@ public class BlockEditorScreen: ScreenObject {
+    }
+
+    @discardableResult
+    public func redoIsDisabled() throws -> BlockEditorScreen {
```
```suggestion
    public func verifyRedoIsDisabled() throws -> BlockEditorScreen {
```
Same as in `undoIsDisabled()`.

geriux replied:
Updated in https://github.com/wordpress-mobile/WordPress-iOS/pull/20930/commits/9747901ba9f5465abdb836197773818f4988ec31

tiagomar commented on WordPress/UITests/Tests/EditorGutenbergTests.swift:
```diff
@@ -55,6 +55,33 @@ class EditorGutenbergTests: XCTestCase {
             .done()
     }
 
+    func testTextPostUndo() throws {
+
+        try BlockEditorScreen()
+            .undoIsDisabled()
+            .enterTextInTitle(text: title)
+            .addParagraphBlock(withText: content)
+            .verifyContentStructure(blocks: 1, words: content.components(separatedBy: " ").count, characters: content.count)
+            .undo()
+            .undo()
+            .verifyContentStructure(blocks: 0, words: 0, characters: 0)
+    }
+
+    func testTextPostRedo() throws {
+
+        try BlockEditorScreen()
+            .redoIsDisabled()
+            .enterTextInTitle(text: title)
+            .addParagraphBlock(withText: content)
+            .verifyContentStructure(blocks: 1, words: content.components(separatedBy: " ").count, characters: content.count)
+            .undo()
+            .undo()
+            .verifyContentStructure(blocks: 0, words: 0, characters: 0)
+            .redo()
+            .redo()
+            .verifyContentStructure(blocks: 1, words: content.components(separatedBy: " ").count, characters: content.count)
+    }
+
```
Here we have a considerable level of duplication, `testTextPostRedo()` actually covers both tests, we would only need to add the `.undoIsDisabled()` check at the beginning. I think this is something we need to change.

Also, name wise, I'm not sure `TextPost` plays an important role here, WDYT about `testUndoRedo()`? It's totally up to you. =]

```suggestion
    func testUndoRedo() throws {

        try BlockEditorScreen()
            .undoIsDisabled()
            .redoIsDisabled()
            .enterTextInTitle(text: title)
            .addParagraphBlock(withText: content)
            .verifyContentStructure(blocks: 1, words: content.components(separatedBy: " ").count, characters: content.count)
            .undo()
            .undo()
            .verifyContentStructure(blocks: 0, words: 0, characters: 0)
            .redo()
            .redo()
            .verifyContentStructure(blocks: 1, words: content.components(separatedBy: " ").count, characters: content.count)
    }

```

geriux replied:
I agree! Thanks for the suggestion about unifying them and the name change, I've applied them in https://github.com/wordpress-mobile/WordPress-iOS/pull/20930/commits/33717fb3409d8c700867a9d6371441dda11b14c6

tiagomar reviewed (changes requested):
Hey @geriux, let me start by saying that I only reviewed the UI Tests related changes, so a review from other engineer is still needed.

I've left a few comments and suggestions, being the one around avoiding duplication, and the naming convention for assertion steps, the ones that require changes. [Don't simply trust the suggestions, I haven't tested them and some do impact others. 😬 ]

Thanks for adding the UI Tests! 🙇‍♂️

momo-ozawa commented on WordPress/Classes/ViewRelated/Post/PostEditorNavigationBarManager.swift:
```diff
@@ -164,7 +255,7 @@ class PostEditorNavigationBarManager {
     }
 
     func reloadBlogTitleView(text: String) {
-        blogTitleViewLabel.text = text
+        blogTitleViewLabel.text = ""
```
@geriux Is this an intentional change?

geriux replied:
Hey @momo-ozawa 👋  Nice catch! With the new design changes we are not displaying the Blog's title in the navigation bar anymore but it was being used by the Aztec editor since they shared the same navigation bar. When I looked into this I realized I needed to do some extra tweaks added in https://github.com/wordpress-mobile/WordPress-iOS/pull/20930/commits/d22dc572bac95ae7b7196903223c7c097bfde21b

## Changing the "Generating Preview" behavior

Currently, we were updating the navigation bar to display when the post preview was being generated, with the new changes there's not much space available so I updated it to be consistent with Android so now it shows like this:

iOS|Android
-|-
<video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/77625f83-d555-421b-ba2f-91f34cbbde64"/> | <video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/732ade80-a84c-4fdf-8e60-c32e1a9622a5" />

## Keeping the old `rightButtonItems` for the Aztec editor

Since the new undo/redo functionality is only available for the Gutenberg editor these shouldn't show for the Aztec editor so I've updated it so it looks like this:

Gutenberg|Aztec
-|-
<kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/87cf331d-1e5e-4b34-bb21-2b0daeca90b3" width="200" /></kbd> | <kbd><img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4885740/a5ea4472-6c7b-45a5-8034-d33151aac039" width="200" /></kbd>

## Removal of the `reloadBlogTitleView` logic

Since the title is not shown in either Gutenberg or Aztec editors, I've removed it.

Let me know what you think! Thanks for the review and testing!

momo-ozawa replied:
Makes sense to me! Thank you!

momo-ozawa reviewed (commented):
Works as described! Left one q

Commit:
BlockEditorScreen - Remove assertion within undo/redo

Commit:
EditorGutenberg tests - Update naming for asserting undo/redo buttons being disabled

Commit:
EditorGutenbergTests - Unify undo/redo tests into one

Commit:
Update PostEditorNavigationBarManager to have two different buttons for both Aztec and Gutenberg, for Gutenberg it will include new undo/redo buttons, for Aztec it will keep the current ones.

It also updates the "Generating preview" text that used to be shown in the header title, now it will show in a SVProgressHUD.

Removes the reloadBlogTitleView code since it is no longer needed.

momo-ozawa reviewed (approved):
LGTM!

tiagomar reviewed (approved):
Thanks for considering the suggestions! 🙇 
Everything looks good from a UI Tests perspective.

Commit:
Merge branch 'trunk' into gutenberg/move-undo-redo-buttons

# Conflicts:
#	Gutenberg/version.rb
#	Podfile.lock
#	WordPress/UITestsFoundation/Screens/Editor/BlockEditorScreen.swift

Commit:
Update Gutenberg ref

Commit:
Update PostEditorNavigationBar close button margin with new values from the editor design changes

Commit:
Update Gutenberg ref

Commit:
Update Gutenberg ref

Commit:
Merge branch 'trunk' into gutenberg/move-undo-redo-buttons

# Conflicts:
#	Podfile.lock

Commit:
Update Podfile.lock

Commit:
Update Release notes
